activerecord/lib/active_record/migration.rb

```diff
@@ -951,15 +952,24 @@ def migrations_paths
         Array(@migrations_paths)
       end
 
+      def match_to_migration_filename?(filename) # :nodoc:
+        filename =~ Migration::MigrationFilenameRegexp
+      end
+
+      def parse_migration_filename(filename) # :nodoc:
+        filename.scan(Migration::MigrationFilenameRegexp).first
+      end
+
       def migrations(paths)
         paths = Array(paths)
 
         files = Dir[*paths.map { |p| "#{p}/**/[0-9]*_*.rb" }]
 
         migrations = files.map do |file|
-          version, name, scope = file.scan(/([0-9]+)_([_a-z0-9]*)\.?([_a-z0-9]*)?\.rb\z/).first
+          basename = File.basename(file)
+          raise IllegalMigrationNameError.new(file) unless match_to_migration_filename?(basename)
 
-          raise IllegalMigrationNameError.new(file) unless version
+          version, name, scope = parse_migration_filename(basename)
```

Comment:
> We could also push the `File.basename` conversions down into the two methods

Yes, I will.

> I think we can still do the old way of checking the version and just have the one regexp comparison.

Sure in old way have just the one comparison, but I think old way was ambiguous. Why `IllegalMigrationNameError` is raised when `version` is not got? In this case I think using `match_to_migration_filename?` makes clear why `IllegalMigrationNameError` is raised.

How do you think about this?